csfieldguide/static/interactives/city-trip/js/city-trip.js

```diff
@@ -0,0 +1,642 @@
+        }
+      }
+      // swap has occurred ending the for-loop. Simulate increment of the for-loop counter.
+      c[i] += 1;
+      // simulate recursive call reaching the base case by bringing the pointer to the base case analog in the array
+      i = 0;
+    } else {
+      // Reset the state and simulate popping the stack by incrementing the pointer
+      c[i] = 0;
+      i += 1;
+    }
+  }
+  if (stoppedMidExecution == false) {
+    // Making sure we have iterated over all paths before showing complete message
+    updateStatus(STATUS_COMPLETE, 'status-complete');
+    $('#start').removeClass('d-none');
+    $('#add-city').removeClass('d-none');
+    $('#remove-city').removeClass('d-none');
+    $('#generate-map').removeClass('d-none');
+    $('#stop').addClass('d-none');
+  }
+  stopPathFinding = true;
+}
+
+
+/** Pass in 2 positions and use (x,y) coords to calc distance. */
+function distanceBetweenCities(edgeStartPos, edgeEndPos) {
+  var a = Math.abs(edgeStartPos.x - edgeEndPos.x);
+  var b = Math.abs(edgeStartPos.y - edgeEndPos.y);
+  var c = Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));
+
+  return c;
+}
+
+
+/** Returns total distance of the path. */
+function getPathDistance(edges) {
+  var distance = 0;
+  for (var i = 0; i < edges.length; i++) {
+    var edge = edges[i];
+    var sourceNode = edge.source();
+    var targetNode = edge.target();
+    distance += distanceBetweenCities(sourceNode.position(), targetNode.position());
+  }
+
+  distance = Mathjs.divide(distance, 100);
+
+  return Math.round(distance);
+}
+
+
+/** Formats the time into years,months,days,hours,minutes,seconds string. */
+function showTimeUnit(unit, value) {
+  unitPlural = unit + 's';
+  unitElement = $('#num-' + unitPlural);
+  if (value > 0 || unit == 'second') {
+    var format = ngettext('1 ' + unit, '%(value)s ' + unitPlural + ' ', value);
+    var timeUnitText = interpolate(format, {"value": value.toLocaleString()}, true);
+    unitElement.html(timeUnitText);
+    unitElement.removeClass('d-none');
+  } else {
+    unitElement.addClass('d-none');
+  }
+}
+
+
+/** Calculates how many years,months,days.. etc there are in runningTimeLeft.
+ *  runningTimeLeft is given in seconds. */
+function formatTime(runningTimeLeft) {
+  var years = Math.floor(runningTimeLeft / 31536000);
+  var months = Math.floor((runningTimeLeft % 31536000) / 2628000);
+  var days = Math.floor((runningTimeLeft % 2628000) / 86400);
+  var hours = Math.floor((runningTimeLeft % 86400) / 3600);
+  var minutes = Math.floor((runningTimeLeft % 3600) / 60);
+  var seconds = (runningTimeLeft % 60).toFixed(1);
+
+  showTimeUnit('year', years);
+  showTimeUnit('month', months);
+  showTimeUnit('day', days);
+  showTimeUnit('hour', hours);
+  showTimeUnit('minute', minutes);
+  showTimeUnit('second', seconds);
+}
+
+
+/** Updates the timer every second and stops the timer if time is up or path finding is stopped. */
+function startTimer(seconds) {
```

Comment:
The timer should probably stop after reaching zero, or at least show zero at the end